server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,590 @@
+
+        for dir_path in direct_entries:
+            if (
+                dir_path.is_dir()
+                and not dir_path.name.endswith(".unpack")
+                and not len(os.listdir(dir_path)) == 0
+            ):
+                tarball_dirs.append(dir_path.name)
+            elif (
+                dir_path.is_dir()
+                and not dir_path.name.endswith(".unpack")
+                and len(os.listdir(dir_path)) == 0
+            ):
+                ihier.add_unexpected_controllers(
+                    Hierarchy.EMPTY_TARBALL_DIRS, controller, dir_path.name
+                )
+            elif (
+                dir_path.is_dir()
+                and dir_path.name.endswith(".unpack")
+                and len(os.listdir(dir_path)) == 0
+            ):
+                unpacking_tarball_dirs.append(dir_path.name)
+            elif dir_path.is_symlink():
+                ihier.add_unexpected_controllers(
+                    Hierarchy.TARBALL_LINKS, controller, dir_path.name
+                )
+
+        if tarball_dirs:
+            verify_tar_dirs(
+                ihier, tarball_dirs, ihier.add_unexpected_controllers, controller
+            )
+
+        if unpacking_tarball_dirs:
+            verify_tar_dirs(
+                ihier,
+                unpacking_tarball_dirs,
+                ihier.add_unexpected_controllers,
+                controller,
+            )
+
+    return 0
+
+
+# Results Hierarchy
+def verify_user_arg(rhier, mdlogcfg, controller, path):
+    """fetch user from the config file and verify its validity"""
+
+    if rhier.path.parent == rhier.config.USERS:
+        """We are reviewing a user tree, so check the user in
+        the configuration.  Version 002 agents use the
+        metadata log to store a user as well.
+        """
+        user_arg = rhier.path.name
+        user = ""
+        try:
+            user = mdlogcfg.get("run", "user")
+        except NoSectionError:
+            pass
+        except NoOptionError:
+            pass
+        if not user:
+            """No user in the metadata.log of the tar ball, but
+            we are examining a link in the user tree that
+            does not have a configured user, report it.
+            """
+            rhier.add_unexpected_controllers(
+                Hierarchy.UNEXPECTED_USER_LINKS, controller, path
+            )
+        elif user_arg != user:
+            """Configured user does not match the user tree in
+            which we found the link."""
+            rhier.add_unexpected_controllers(
+                Hierarchy.WRONG_USER_LINKS, controller, path
+            )
+
+    return 0
+
+
+def path_below_controller(dir_path, controller):
+
+    controller_path = os.path.join(controller, "")
+    path = dir_path.split(controller_path, 1)[1]
+
+    return Path(path)
+
+
+def list_direct_entries(hierarchy, controller):
+
+    direct_entries = list()
+    for root, dirs, files in os.walk(Path(hierarchy, controller), topdown=True):
```

Comment:
Add the `followlinks=True` argument, even though it is the default: that makes it clear that we are *not* following links, without having to check the documentation of `os.walk()`.

Reply:
this seems to be creating issues

Reply:
this seems to be creating issues

Reply:
Showing some invalid Tarball links in results hierarchy